### PR TITLE
feat(space-wizard): added forge integration

### PR DESCRIPTION
**Pull request by mitchharpur**
reworked forge integration pull request (basically same as previous) but included api locator and auth token submission. mostly new files with new functionality

-wizard now uses a workflow object to orchestrate transitions. This allow for strong decoupling of wizard screen to components, but allow next,prev,finish actions to be called from seperate components

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: bbec2929eaccedde0c0cbd6d2cd9adb883187b1d
  * Your commit message does not contain a body

Guidelines are available at https://github.com/fabric8io/fabric8-ui

---
This message was auto-generated by https://gitcop.com

**Comment by joshuawilson** [comment]
[test]

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: bbec2929eaccedde0c0cbd6d2cd9adb883187b1d
  * Your commit message does not contain a body

 * Commit: 3f4c7e6c65e66034f2f6cba304ccbc1f46595f42
  * Your commit message does not contain a body
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 424a231ce597bf8dae0b07711b5136e8074396df
  * Your commit message does not contain a body

 * Commit: 6b9dd577488eda1156cd112e27e3bc24ba8cb0cd
  * Your commit message does not contain a body

 * Commit: 8cefe535eee24dfcc46f325a1ba62ae0b466bc15
  * Your commit message does not contain a body
  * Your subject line is longer than 50 characters

 * Commit: 47fcd149abdabe55ebe72f34f92eae2f2a6acbee
  * Your commit message does not contain a body
  * Your subject line is longer than 50 characters

Guidelines are available at https://github.com/fabric8io/fabric8-ui

---
This message was auto-generated by https://gitcop.com

**Comment by joshuawilson** [comment]
[test]

**Comment by joshuawilson** [comment]
@mitchharpur Please rebase again, there are merge conflicts. Please squash/fixup the merge commits.

File: config/webpack.prod.js [hunk 1]
```diff
@@ -29,6 +29,7 @@ const OfflinePlugin = require('offline-plugin');
  * Webpack Constants
  */
 const ENV = process.env.NODE_ENV = process.env.ENV = 'production';
+const FABRIC8_FORGE_API_URL = process.env.FABRIC8_FORGE_API_URL || 'https://forge.api.prod-preview.openshift.io';
```
Comment (joshuawilson) [inline config/webpack.prod.js]: I think this should be configured in the docker container or somewhere else. 

@pmuir where did you put the other URLs?

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 99faa01dddab8707d2db3124699c586c15a3cd4e
  * Your commit message does not contain a body

Guidelines are available at https://github.com/fabric8io/fabric8-ui

---
This message was auto-generated by https://gitcop.com

File: config/webpack.prod.js (hunk 1 above)
Comment (pmuir) [inline config/webpack.prod.js]: You need to change 

```typescript
   readonly DEFAULT_API_PREFIXES = new Map<string, string>([
      ['wit', 'api'],
      ['recommender', 'recommender'],
     ['sso', 'sso'],
     ['forge', 'forge']
    ]);
```

to

```typescript
   readonly DEFAULT_API_PREFIXES = new Map<string, string>([
     ['wit', 'api'],
     ['recommender', 'recommender'],
     ['sso', 'sso'],
     ['forge', 'forge.api']
    ]);
```

And then you can remove the override from the prod webpack. The api-locator will automatically configure the right URL for you.

File: package.json [hunk 2]
```diff
@@ -101,6 +101,7 @@
     "@angularclass/conventions-loader": "1.0.13",
     "@angularclass/hmr": "1.2.2",
     "@angularclass/hmr-loader": "3.0.2",
+    "angular-2-dropdown-multiselect": "^1.0.5",
```
Comment (joshuawilson) [inline package.json]: remove auto update ^

File: src/app/space-wizard/common/logger.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+}
+/** 
+ * This is just a quick and dirty functional style logger allowing for unqique logger 'instances'
+ * for each class 
+ */
+
+@Injectable()
+export class LoggerFactory {
+  constructor(){
+    console.log("%cLoggerFactory: %cNew instace.",this.styles.origin,this.styles.message)
```
Comment (joshuawilson) [inline src/app/space-wizard/common/logger.ts]: instace is mis-spelled

File: src/app/space-wizard/common/logger.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  }
+  styles={
+    origin:"background:linear-gradient(#444, #333); border-radius:15px;padding:4px;color:lime;font-style:italic;border-left:solid 0px orangered;padding:3px;padding-left:10px;padding-right:10px",
+    instance: "background:linear-gradient(#444, #333);color:orangered; border-radius:10px;padding:3px;margin:3px 0;",
+    message: "background:linear-gradient(#444, #333);color:white; border-radius:10px;padding:3px 10px;",
+
+  }
```
Comment (joshuawilson) [inline src/app/space-wizard/common/logger.ts]: should end in ;

File: src/app/space-wizard/common/logger.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        method = "error";
+      }
+      if (entry.warning === true) {
+        method = "warn";
+      }
+      if (entry.info === true) {
+        method = "info";
+      }
+      console[method].apply(null, [`%c${origin}%c ${instance} %c${entry.message || ""}`, me.styles.origin,me.styles.instance, me.styles.message]);
+    };
```
Comment (joshuawilson) [inline src/app/space-wizard/common/logger.ts]: semicolon not needed

File: src/app/space-wizard/common/logger.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    function loggerDelegate(options: string | ILogEntry) {
+      let entry = { message: "" };
+      if (typeof options === "string") {
+        entry.message = options || "";
+      }
+      else {
+        Object.assign(entry, options);
+      }
+      addLogEntry(entry);
+    };
```
Comment (joshuawilson) [inline src/app/space-wizard/common/logger.ts]: semicolon not needed

File: src/app/space-wizard/components/wizard-dynamic-step/wizard-dynamic-step.component.html [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+<section style="overflow:auto;" class="flex-container in-column-direction flex-grow-1 flex-align-content-stretch margin-top-15 container-fluid width-100">
+  <div class="col-xs-12">
+    <div style="min-height:50vh;"  class="form-horizontal">
+      <div  class="form-group" *ngFor="let field of fieldSet">
+        <label class="control-label col-xs-4">{{field.display.label}}</label>
+        <div class="col-xs-8"  [ngSwitch]="field.display.inputType">
+            <div *ngSwitchCase="'singleInput'" [ngSwitch]="field.valueType">
+              <input *ngSwitchCase="'string'" name="{{field.name}}" [(ngModel)]="field.value" type="text" class="form-control" placeholder="{{field.description}}"
+                [required]="field.display.required" [disabled]="!field.display.enabled" />
+              <input *ngSwitchCase="'boolean'" type="checkbox" name="{{field.name}}" [(ngModel)]="field.value" class="form-control" [required]="field.display.required"
```
Comment (joshuawilson) [inline src/app/space-wizard/components/wizard-dynamic-step/wizard-dynamic-step.component.html]: there is no associated label for inputs

File: src/app/space-wizard/components/wizard-dynamic-step/wizard-dynamic-step.component.ts [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+  providers: [IAppGeneratorServiceProvider.FactoryProvider]
+})
+export class WizardDynamicStepComponent implements OnInit, OnDestroy, OnChanges {
+
+  // keep track of the number of instances
+  static instanceCount: number = 1;
+  /** logger delegate delegates logging to a logger */
+  private log: ILoggerDelegate = () => { };
+
+  private _workflow: IWorkflow
```
Comment (joshuawilson) [inline src/app/space-wizard/components/wizard-dynamic-step/wizard-dynamic-step.component.ts]: needs a semicolon

File: src/app/space-wizard/components/wizard-dynamic-step/wizard-dynamic-step.component.ts [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+  ngOnInit() {
+    this.log(`ngOnInit ...`)
+  }
+
+  ngOnDestroy() {
+    this.log(`ngOnDestory ...`)
+  }
+  /** handle all changes to @Input properties */
+  ngOnChanges(changes: SimpleChanges) {
+    this.log(`ngOnChanges ...`)
```
Comment (joshuawilson) [inline src/app/space-wizard/components/wizard-dynamic-step/wizard-dynamic-step.component.ts]: semicolon

File: src/app/space-wizard/components/wizard-dynamic-step/wizard-dynamic-step.component.ts [hunk 10]
```diff
@@ -0,0 +1,231 @@
+import { Component, OnInit, OnChanges, OnDestroy, SimpleChanges, SimpleChange, Input, Output, Inject, EventEmitter } from '@angular/core';
```
Comment (joshuawilson) [inline src/app/space-wizard/components/wizard-dynamic-step/wizard-dynamic-step.component.ts]: SimpleChange, Output, EventEmitter not used

File: src/app/space-wizard/components/wizard-dynamic-step/wizard-dynamic-step.component.ts [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+import { Router } from '@angular/router';
```
Comment (joshuawilson) [inline src/app/space-wizard/components/wizard-dynamic-step/wizard-dynamic-step.component.ts]: not used

File: src/app/space-wizard/components/wizard-dynamic-step/wizard-dynamic-step.component.ts [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+
+import { IWorkflowStep, IWorkflowTransition, IWorkflow, WorkflowTransitionDirection } from '../../models/workflow';
```
Comment (joshuawilson) [inline src/app/space-wizard/components/wizard-dynamic-step/wizard-dynamic-step.component.ts]: IWorkflowStep not used

File: src/app/space-wizard/components/wizard-dynamic-step/wizard-dynamic-step.component.ts [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+import { LoggerFactory, ILoggerDelegate } from '../../common/logger';
+import { INotifyPropertyChanged } from '../../core/component'
+
+import { IFieldInfo, IFieldSet, IAppGeneratorResponse, IAppGeneratorRequest, IAppGeneratorCommand, IAppGeneratorService, IAppGeneratorServiceProvider } from '../../services/app-generator.service';
```
Comment (joshuawilson) [inline src/app/space-wizard/components/wizard-dynamic-step/wizard-dynamic-step.component.ts]: IFieldInfo and IAppGeneratorCommand import not used

File: src/app/space-wizard/components/wizard-dynamic-step/wizard-dynamic-step.component.ts [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+
+
+import { IMultiSelectOption, IMultiSelectSettings, IMultiSelectTexts } from 'angular-2-dropdown-multiselect';
```
Comment (joshuawilson) [inline src/app/space-wizard/components/wizard-dynamic-step/wizard-dynamic-step.component.ts]: IMultiSelectOption import not used

File: src/app/space-wizard/models/concrete/workflow-transition.ts [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+  constructor(options: Partial<IWorkflowTransition> = { from: null, to: null, continue: true, direction: WorkflowTransitionDirection.GO, context: {} }) {
+    this.context = {};
+    this.continue = true;
+    this.direction = WorkflowTransitionDirection.GO;
+    Object.assign(this, options);
+  }
+  from?: IWorkflowStep;
+  to?: IWorkflowStep;
+  context: IWorkflowTransitionContext
+  continue: boolean = false;
```
Comment (joshuawilson) [inline src/app/space-wizard/models/concrete/workflow-transition.ts]: continue is a reserved word

File: src/app/space-wizard/models/contracts/forge-input.ts [hunk 16]
```diff
@@ -0,0 +1,18 @@
+import {IForgeValueChoice} from './forge-value-choice';
+
+export interface IForgeInput {
+  name: string;
+  label: string;
+  description?: string;
+  note?:string;
+  class: string;
```
Comment (joshuawilson) [inline src/app/space-wizard/models/contracts/forge-input.ts]: class is reserved

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 99faa01dddab8707d2db3124699c586c15a3cd4e
  * Your commit message does not contain a body

 * Commit: b951e29b5137674ffc7611044336a963ddecc612
  * Your commit message does not contain a body
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**
  * Your subject line is longer than 50 characters

Guidelines are available at https://github.com/fabric8io/fabric8-ui

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 99faa01dddab8707d2db3124699c586c15a3cd4e
  * Your commit message does not contain a body

 * Commit: b951e29b5137674ffc7611044336a963ddecc612
  * Your commit message does not contain a body
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**
  * Your subject line is longer than 50 characters

 * Commit: 64b344e4b0418d081ecdea155ae171436e0969b9
  * Your commit message does not contain a body
  * Your subject line is longer than 50 characters

Guidelines are available at https://github.com/fabric8io/fabric8-ui

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 99faa01dddab8707d2db3124699c586c15a3cd4e
  * Your commit message does not contain a body

 * Commit: b951e29b5137674ffc7611044336a963ddecc612
  * Your commit message does not contain a body
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**
  * Your subject line is longer than 50 characters

 * Commit: 64b344e4b0418d081ecdea155ae171436e0969b9
  * Your commit message does not contain a body
  * Your subject line is longer than 50 characters

 * Commit: cc552ec147d5015d6f3c0067653da8a6c50c03c4
  * Your commit message does not contain a body
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/fabric8io/fabric8-ui

---
This message was auto-generated by https://gitcop.com

**Comment by joshuawilson** [comment]
[test]

File: config/webpack.prod.js (hunk 1 above)
Comment (mitchharpur) [inline config/webpack.prod.js]: done :can you approve the fix

Comment (pmuir) [inline config/webpack.prod.js]: You still need to remove the default value from webpack.prod.js

File: src/app/shared/api-locator.service.ts [hunk 17]
```diff
@@ -7,15 +7,17 @@ export class ApiLocatorService {
...
 
     ]
   );
 
   readonly DEFAULT_API_PREFIXES = new Map<string, string>([
     ['wit', 'api'],
     ['recommender', 'recommender'],
-    ['sso', 'sso']
+    ['sso', 'sso'],
+    ['forge', 'forge.api']
```
Comment (pmuir) [inline src/app/shared/api-locator.service.ts]: Looks good.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 99faa01dddab8707d2db3124699c586c15a3cd4e
  * Your commit message does not contain a body

 * Commit: b951e29b5137674ffc7611044336a963ddecc612
  * Your commit message does not contain a body
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**
  * Your subject line is longer than 50 characters

 * Commit: 64b344e4b0418d081ecdea155ae171436e0969b9
  * Your commit message does not contain a body
  * Your subject line is longer than 50 characters

 * Commit: cc552ec147d5015d6f3c0067653da8a6c50c03c4
  * Your commit message does not contain a body
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 0116829050786133def17ed031c46da743225085
  * Your commit message does not contain a body
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/fabric8io/fabric8-ui

---
This message was auto-generated by https://gitcop.com

**Comment by joshuawilson** [comment]
[test]

**Pull request merged**